File: ios/Classes/Offline/OfflineController.swift
import MapboxMaps
import Flutter

final class OfflineController: _OfflineManager {
    enum `Error`: Swift.Error {
        case invalidStyleURI
        case invalidStylePackLoadOptions
    }

    private lazy var offlineManager = MapboxCoreMaps.OfflineManager()
    private var progressHandlers: [String: AnyFlutterStreamHandler] = [:]
    private let messenger: SuffixBinaryMessenger

    init(messenger: SuffixBinaryMessenger) {
        self.messenger = messenger
    }

    func loadStylePack(
        styleURI uri: String,
        loadOptions: StylePackLoadOptions,
        completion: @escaping (Result<StylePack, Swift.Error>) -> Void
    ) {
        guard let styleURI = StyleURI(rawValue: uri) else {
            completion(.failure(Error.invalidStyleURI))
            return
        }
        guard let loadOptions = MapboxCoreMaps.StylePackLoadOptions(fltValue: loadOptions) else {
            completion(.failure(Error.invalidStylePackLoadOptions))
            return
        }

        offlineManager.loadStylePack(
            for: styleURI,
            loadOptions: loadOptions) { [weak self] progress in
                self?.progressHandlers[uri]?.eventSink?(progress.toFLTStylePackLoadProgress().toList())
            } completion: { [weak self] result in
                executeOnMainThread(completion)(result.map { $0.toFLTStylePack() })
                self?.progressHandlers.removeValue(forKey: uri)
            }
    }

    func addStylePackLoadProgressListener(styleURI: String) {
        let handler = AnyFlutterStreamHandler()
        let eventChannel = FlutterEventChannel(name: "com.mapbox.maps.flutter/\(messenger.suffix)/\(styleURI)", binaryMessenger: messenger.messenger)
        eventChannel.setStreamHandler(handler)
        progressHandlers[styleURI] = handler
    }

    func removeStylePack(styleURI: String, completion: @escaping (Result<StylePack, Swift.Error>) -> Void) {
        guard let styleURI = StyleURI(rawValue: styleURI) else {
            completion(.failure(Error.invalidStyleURI))
            return
        }
        offlineManager.removeStylePack(for: styleURI) { result in
            executeOnMainThread(completion)(result.map { $0.toFLTStylePack() })
        }
    }

    func stylePack(styleURI: String, completion: @escaping (Result<StylePack, Swift.Error>) -> Void) {
        guard let styleURI = StyleURI(rawValue: styleURI) else {
            completion(.failure(Error.invalidStyleURI))
            return
        }
        offlineManager.stylePack(for: styleURI) { result in
            executeOnMainThread(completion)(result.map { $0.toFLTStylePack() })
        }
    }

    func stylePackMetadata(styleURI: String, completion: @escaping (Result<[String: Any], Swift.Error>) -> Void) {
        guard let styleURI = StyleURI(rawValue: styleURI) else {
            completion(.failure(Error.invalidStyleURI))
            return
        }
        offlineManager.stylePackMetadata(for: styleURI) { result in
            executeOnMainThread(completion)(result.map { $0 as? [String: Any] ?? [:] })
        }
    }

    func allStylePacks(completion: @escaping (Result<[StylePack], Swift.Error>) -> Void) {
        offlineManager.allStylePacks { result in
            executeOnMainThread(completion)(result.mapElement { $0.toFLTStylePack() })
        }
    }
}

extension OfflineSwitch: _OfflineSwitch {
    func setMapboxStackConnected(connected: Bool) throws {
        isMapboxStackConnected = connected
    }

    func isMapboxStackConnected() throws -> Bool {
        return isMapboxStackConnected
    }

    func reset() throws {}
}

File: android/src/main/kotlin/com/mapbox/maps/mapbox_maps/offline/OfflineController.kt
package com.mapbox.maps.mapbox_maps.offline

import android.content.Context
import android.os.Handler
import com.mapbox.maps.OfflineManager
import com.mapbox.maps.mapbox_maps.pigeons.*
import com.mapbox.maps.mapbox_maps.toFLTStylePack
import com.mapbox.maps.mapbox_maps.toFLTStylePackLoadProgress
import com.mapbox.maps.mapbox_maps.toFLTValue
import com.mapbox.maps.mapbox_maps.toResult
import com.mapbox.maps.mapbox_maps.toStylePackLoadOptions
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.EventChannel.StreamHandler

private const val EVENT_CHANNEL_PREFIX = "com.mapbox.maps.flutter/offline"

class OfflineController(
  private val context: Context,
  private val messenger: BinaryMessenger
) : _OfflineManager {

  private val offlineManager = OfflineManager()
  private var progressHandlers = HashMap<String, EventChannel.EventSink>()
  private val mainHandler = Handler(context.mainLooper)

  override fun loadStylePack(
    styleURI: String,
    loadOptions: StylePackLoadOptions,
    callback: (Result<StylePack>) -> Unit
  ) {
    offlineManager.loadStylePack(
      styleURI,
      loadOptions.toStylePackLoadOptions(),
      { progress ->
        mainHandler.post {
          progressHandlers[styleURI]?.success(progress.toFLTStylePackLoadProgress().toList())
        }
      },
      { expected ->
        mainHandler.post {
          callback(expected.toResult { it.toFLTStylePack() })
          progressHandlers.remove(styleURI)?.endOfStream()
        }
      }
    )
  }

  override fun removeStylePack(styleURI: String, callback: (Result<StylePack>) -> Unit) {
    offlineManager.removeStylePack(
      styleURI
    ) { expected ->
      mainHandler.post {
        callback(expected.toResult { it.toFLTStylePack() })
      }
    }
  }

  override fun addStylePackLoadProgressListener(styleURI: String) {
    val eventChannel = EventChannel(messenger, "com.mapbox.maps.flutter/offline/$styleURI")
    eventChannel.setStreamHandler(
      object : StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
          events?.let { progressHandlers[styleURI] = it }
        }

        override fun onCancel(arguments: Any?) {
          progressHandlers.remove(styleURI)
        }
      }
    )
  }

  override fun stylePack(styleURI: String, callback: (Result<StylePack>) -> Unit) {
    offlineManager.getStylePack(styleURI) { expected ->
      mainHandler.post {
        callback(expected.toResult { it.toFLTStylePack() })
      }
    }
  }

  override fun stylePackMetadata(styleURI: String, callback: (Result<Map<String, Any>>) -> Unit) {
    offlineManager.getStylePackMetadata(styleURI) { expected ->
      mainHandler.post {
        callback(expected.toResult { it.toFLTValue() as? kotlin.collections.Map<kotlin.String, kotlin.Any> ?: kotlin.collections.emptyMap() })
      }
    }
  }

  override fun allStylePacks(callback: (Result<List<StylePack>>) -> Unit) {
    offlineManager.getAllStylePacks { expected ->
      mainHandler.post {
        callback(expected.toResult { it.map { stylePack -> stylePack.toFLTStylePack() } })
      }
    }
  }
}